libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -761,6 +761,33 @@ void GCS_MAVLINK::handle_radio_status(mavlink_message_t *msg, DataFlash_Class &d
     }
 }
 
+/*
+  return a timesync request
+  Sends back ts1 as received, and tc1 is the local timestamp in usec
+ */
+void GCS_MAVLINK::handle_timesync(mavlink_message_t *msg)
+{
+    // Decode incoming timesync message
+    mavlink_timesync_t tsync;
+    mavlink_msg_timesync_decode(msg, &tsync);
+
+    // Create new timestruct to return
+    mavlink_timesync_t rsync;
+    rsync.ts1 = tsync.ts1;
+    rsync.tc1 = AP_HAL::micros64();
+
+    // If message is sent with tc > 1, calculate the offset and return that instead
+    if (tsync.tc1 > 0) {
+        rsync.tc1 = tsync.tc1 - AP_HAL::micros64();
```

Comment:
I don't believe this is correct, tc1 should be the absolute time, not a delta. this could also lead to a timesync flood.
tsync.ts1 = time at send
tsync.tc1 = 0 > time at client (unknown)
rsync.ts1 = copy of tsync.ts1 (time of the origin)
rsync.tx1 = out local absolute time

if tsync.tc1 > 0, then this is a time reply that originated from ourself, so there is no need to resend the message at all. plus we now have the info of the local time we sent the request (stored as we asked for it), and the time we acquired the response. (ie calc the offset)
please refer to here https://pixhawk.ethz.ch/docs/mavlink__receiver_8cpp_source.html